Initial contribution of the device firmware update API based on the d…

tomhoefer opened the pull request:
…iscussions of issue #1153

Extended the core.thing package with APIs in order to update the firmware of a thing. Binding developers have to implement the interface FirmwareUpdateHandler in order that the firmware of their things can be updated.

The FirmwareRegistry is responsible to track all FirmwareProviders and to make the firmware available to the system. The FirmwareUpdateService is responsible to track all FirmwareUpdateHandlers and both to provide the firmware status and to start the firmware update process.

Signed-off-by: Thomas Höfer <t.hoefer@telekom.de>

maggu2810 commented:
Hi @tomhoefer, will there be at least one public Eclipse SmartHome binding that is using that feature? Would be nice to test this and see this running.

tomhoefer commented:
We are going to implement this feature for the Hue Binding. However for this binding we need some "special" firmware provider because the firmware upgrade for the Hue Bridge is a simple REST call against the bridge instead of using a real binary file... Of course we have to extend the Jue library for this purpose...

Furthermore there will be a new ESH binding in the future for which we originally developed this feature...

hmerk commented:
It could be usefull for the WeMo Binding as well, as there are firmware update actions described in the UPnP implementation.

sjka commented:
I've been running into trouble with the tests. As it seems, they are a bit flaky currently:

```
16:48:55 assert that firmware update is rejected for things whose handler is not a firmware update handler(org.eclipse.smarthome.core.thing.firmware.FirmwareUpdateServiceOSGiTest)  Time elapsed: 0.201 sec
16:48:55 assert that firmware update handlers are ignored that are not thing handlers(org.eclipse.smarthome.core.thing.firmware.FirmwareUpdateServiceOSGiTest)  Time elapsed: 0.122 sec
16:48:55 assert that firmware update is rejected if firmware does not exist(org.eclipse.smarthome.core.thing.firmware.FirmwareUpdateServiceOSGiTest)  Time elapsed: 0.082 sec
16:48:55 assert that update firmware is a safe method call(org.eclipse.smarthome.core.thing.firmware.FirmwareUpdateServiceOSGiTest)  Time elapsed: 0.103 sec  <<< FAILURE!
16:48:55 java.lang.AssertionError: 
16:48:55 Expected: is <UPDATABLE>
16:48:55      but: was <UNKNOWN>
16:48:55        at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
16:48:55        at org.junit.Assert.assertThat(Assert.java:865)
16:48:55        at org.junit.Assert.assertThat(Assert.java:832)
16:48:55        at org.junit.Assert$assertThat.callStatic(Unknown Source)
16:48:55        at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCallStatic(CallSiteArray.java:53)
16:48:55        at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callStatic(AbstractCallSite.java:157)
16:48:55        at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callStatic(AbstractCallSite.java:169)
16:48:55        at org.eclipse.smarthome.core.thing.firmware.FirmwareUpdateServiceOSGiTest.assert that update firmware is a safe method call(FirmwareUpdateServiceOSGiTest.groovy:523)
16:48:55 
16:48:55 assert that firmware update is rejected if firmware is not suitable for thing(org.eclipse.smarthome.core.thing.firmware.FirmwareUpdateServiceOSGiTest)  Time elapsed: 0.069 sec
16:48:55 assert that given thing uid is checked for getFirmwareStatus operation(org.eclipse.smarthome.core.thing.firmware.FirmwareUpdateServiceOSGiTest)  Time elapsed: 0.064 sec
16:48:55 assert that given thing uid is checked for updateFirmware operation(org.eclipse.smarthome.core.thing.firmware.FirmwareUpdateServiceOSGiTest)  Time elapsed: 0.075 sec
16:48:55 assert that firmware upgrade works(org.eclipse.smarthome.core.thing.firmware.FirmwareUpdateServiceOSGiTest)  Time elapsed: 0.318 sec
16:48:55 assert that BaseThingHandlerFactory registers firmware update handler(org.eclipse.smarthome.core.thing.firmware.FirmwareUpdateServiceOSGiTest)  Time elapsed: 0.103 sec
16:48:55 assert that firmware status change is only propagated once(org.eclipse.smarthome.core.thing.firmware.FirmwareUpdateServiceOSGiTest)  Time elapsed: 0.186 sec
16:48:55 assert that prerequisite version is checked(org.eclipse.smarthome.core.thing.firmware.FirmwareUpdateServiceOSGiTest)  Time elapsed: 0.239 sec  <<< ERROR!
16:48:55 java.lang.IllegalArgumentException: Firmware with UID FirmwareUID [thingTypeUID=simpleBinding:simpleThingType1, firmwareVersion=1.1.3] requires at least firmware version 1.1.1-fix to get installed. But the current firmware version of the thing with UID simpleBinding:simpleThingType1:simpleThing2 is 1.1.1.
16:48:55        at org.eclipse.smarthome.core.thing.firmware.FirmwareUpdateService.validateFirmwareSuitability(FirmwareUpdateService.java:214)
16:48:55        at org.eclipse.smarthome.core.thing.firmware.FirmwareUpdateService.validateFirmwareUpdateConditions(FirmwareUpdateService.java:201)
16:48:55        at org.eclipse.smarthome.core.thing.firmware.FirmwareUpdateService.updateFirmware(FirmwareUpdateService.java:164)
16:48:55        at org.eclipse.smarthome.core.thing.firmware.FirmwareUpdateService$updateFirmware.call(Unknown Source)
16:48:55        at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:45)
16:48:55        at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:108)
16:48:55        at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:120)
16:48:55        at org.eclipse.smarthome.core.thing.firmware.FirmwareUpdateServiceOSGiTest.assert that prerequisite version is checked(FirmwareUpdateServiceOSGiTest.groovy:471)
16:48:55 
16:48:55 assert that thing has still current firmware status as long as binary is read(org.eclipse.smarthome.core.thing.firmware.FirmwareUpdateServiceOSGiTest)  Time elapsed: 0.41 sec
16:48:55 assert that firmware status is propagated regularly through job(org.eclipse.smarthome.core.thing.firmware.FirmwareUpdateServiceOSGiTest)  Time elapsed: 0.115 sec  <<< FAILURE!
16:48:55 java.lang.AssertionError: 
16:48:55 Expected: is "1.1.2"
16:48:55      but: was "1.1.1"
16:48:55        at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
16:48:55        at org.junit.Assert.assertThat(Assert.java:865)
16:48:55        at org.junit.Assert.assertThat(Assert.java:832)
16:48:55        at org.junit.Assert$assertThat.callStatic(Unknown Source)
16:48:55        at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCallStatic(CallSiteArray.java:53)
16:48:55        at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callStatic(AbstractCallSite.java:157)
16:48:55        at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callStatic(AbstractCallSite.java:169)
16:48:55        at org.eclipse.smarthome.core.thing.firmware.FirmwareUpdateServiceOSGiTest.assert that firmware status is propagated regularly through job(FirmwareUpdateServiceOSGiTest.groovy:240)
16:48:55 
16:48:55 assert that firmware downgrade works(org.eclipse.smarthome.core.thing.firmware.FirmwareUpdateServiceOSGiTest)  Time elapsed: 0.197 sec
16:48:55 assert that firmware update is rejected if firmware is already updating for thing(org.eclipse.smarthome.core.thing.firmware.FirmwareUpdateServiceOSGiTest)  Time elapsed: 0.05 sec
16:48:55 test that firmware status is changed once a firmware provider is available(org.eclipse.smarthome.core.thing.firmware.FirmwareUpdateServiceOSGiTest)  Time elapsed: 0.171 sec
16:48:55 assert that firmware update is rejected if thing is offline(org.eclipse.smarthome.core.thing.firmware.FirmwareUpdateServiceOSGiTest)  Time elapsed: 0.079 sec
16:48:55 assert that correct firmware status is provided(org.eclipse.smarthome.core.thing.firmware.FirmwareUpdateServiceOSGiTest)  Time elapsed: 0.131 sec  <<< FAILURE!
16:48:55 java.lang.AssertionError: 
16:48:55 Expected: is <CURRENT>
16:48:55      but: was <UPDATABLE>
16:48:55        at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
16:48:55        at org.junit.Assert.assertThat(Assert.java:865)
16:48:55        at org.junit.Assert.assertThat(Assert.java:832)
16:48:55        at org.junit.Assert$assertThat.callStatic(Unknown Source)
16:48:55        at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCallStatic(CallSiteArray.java:53)
16:48:55        at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callStatic(AbstractCallSite.java:157)
16:48:55        at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callStatic(AbstractCallSite.java:169)
16:48:55        at org.eclipse.smarthome.core.thing.firmware.FirmwareUpdateServiceOSGiTest.assert that correct firmware status is provided(FirmwareUpdateServiceOSGiTest.groovy:187)
16:48:55 
16:48:55 assert that get firmware status is rejected if thing does not exist(org.eclipse.smarthome.core.thing.firmware.FirmwareUpdateServiceOSGiTest)  Time elapsed: 0.045 sec
16:48:55 assert that firmware status is unknown if no firmware provider is available(org.eclipse.smarthome.core.thing.firmware.FirmwareUpdateServiceOSGiTest)  Time elapsed: 0.09 sec
16:48:55 assert that firmware update is rejected if thing does not exist(org.eclipse.smarthome.core.thing.firmware.FirmwareUpdateServiceOSGiTest)  Time elapsed: 0.113 sec
16:48:55 assert that given firmware uid is checked for updateFirmware operation(org.eclipse.smarthome.core.thing.firmware.FirmwareUpdateServiceOSGiTest)  Time elapsed: 0.043 sec
16:48:55 Running org.eclipse.smarthome.core.thing.firmware.FirmwareRegistryOSGiTest
16:48:55 Tests run: 8, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.104 sec - in org.eclipse.smarthome.core.thing.firmware.FirmwareRegistryOSGiTest
```

Could you please have a look?

tomhoefer commented:
thanks @sjka 

On my macbook all tests passes. I expect that I have to make them more robust for "slower" machines

maggu2810 commented:
A general question about the timing requirements in our tests:
Why do we use "small" timeouts? A test should succeed on "slower" systems, too. On "faster" systems a bigger timeout doesn't matter, as the test continues if the condition is met. On "slower" system the test could run longer, but should not fail just because the machine is not "fast enough" or has a high CPU load.

WDYT?

tomhoefer commented:
I agree on that...

tomhoefer commented:
Tests are now more robust ....

tomhoefer commented:
Updated this PR:
- core.thing bundle exports now org.eclipse.smarthome.core.thing.firmware package
- made FirmwareRegistry more robust if invocation of firmware provider failed

tomhoefer commented:
At the moment we are mainly considering tow different firmware update processes.
- Standard Firmware Update
- Background OTA Firmware Update

## Standard Firmware Update
For most devices a standard firmware update process is suitable:
- Check if firmware can be updated
- Download firmware
- Transfer firmware to device
- Apply new firmware to device (which may imply a reboot of device)
- Wait until device is up and running again 
- Check if requested FW version is really applied
The FirmwareUpdateService uses the FirmwareUpdateHandler to request the current firmware version of a thing and the firmware registry to get all available firmwares for the thing type of the thing. If the thing is updatable and if there is a later firmware available then the firmware status is UPDATABLE. For the update the FirmwareUpdateService retrieves the Firmware from the FirmwareRegistry and calls updateFirmware() of the FirmwareUpdateHandler. The handler reads the firmware image from the provided InputStream and uploads it to the affected device. Most devices will trigger a reboot automatically after they have updated their firmware. 
![standard_dfu](https://cloud.githubusercontent.com/assets/5469298/14385707/1d5e4c7e-fda2-11e5-87f4-d1b945793dcc.png)

### Background OTA Firmware Update
There are a few devices that support a different approach for firmware update. In this approach the firmware is transferred "over the air" (OTA) in the background to the device without updating the firmware immediately. The firmware will be activated by a separate step later. Battery based devices may even need a manual interaction on device to wakeup the device to be able to enforce the firmware update NOW.
- FirmwareUpdatehandler checks on a regular base if firmware updates are available (e.g. 1x/day)
- If so, download the firmware image
- Transfer firmware to device, device is not yet updatable
- Device is now updatable 
- Apply new firmware to device (which may imply a reboot of device)
- Wait until device is up and running again 
- Check if requested FW version is really applied
The binding must implement some methods to request the firmware directly from the FirmwareRegistry. If it detects newer firmware it automatically transfers this firmware to the device, that needs to be updated. Only after the firmware image has been successfully tranferred to the device, its status changes to updatable. The FirmwareUpdateService can now be triggered to start the firmware update on the device without the need to read the whole firmware image first. 
![background_ota_update](https://cloud.githubusercontent.com/assets/5469298/14385714/25ca4732-fda2-11e5-8f63-3eb5e702e58b.png)

hmerk commented:
@tomhoefer what about devices where you just trigger the update process and the device get's it directly from a cloud service ? That's how it happens with the WeMo devices.

tomhoefer commented:
This is also supported by the current implementation... The difference here is that you have to write your own FirmwareProvider in order to get the firware information from the cloud... I have prototyped that with the Hue bridge....

tomhoefer commented:
I will push some changes regarding device firmware update with the beginning of next week. I got feedback from several developers that I would like to implement. The main changes are:

1. The firmware status enum will consist of the following values:
UNKNOWN, UP_TO_DATE, UPDATE_AVAILABLE, UPDATE_EXECUTBALE

2. The operation isUpdatable is renamed to isUpdateExecutable

3. Background OTA firmware update handlers get the firmware to be transferred to the device injected by the firmware update service.

4. getFirmwareStatus will return domain object FirmwareStatusInfo instead of firmware status enumeration. The FirmwareStatusInfo will consist of FirmwareStatus and the corresponding firmware uid to be updated

5. FirmwareUpdateHandlers have to have the possibility to post their status (READING_FIRMWARE, TRANSFERRING_FIRMWARE, UPDATING_FIRMWARE, UPDATE_SUCCESSFUL, etc)

tomhoefer commented:
Updated this PR... so any feedback is appreciated....

kaikreuzer commented on bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/firmware/FirmwareStatus.java:
```diff
@@ -0,0 +1,41 @@
+    UPDATE_AVAILABLE,
+
+    /** There is a newer firmware of the thing available and the firmware update for the thing can be executed. */
+    UPDATE_EXECUTBALE;
```
typo: EXECUTABLE

tomhoefer replied:
thanks.... I read it a few times but havent seen this

kaikreuzer commented on bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/firmware/ProgressCallbackImpl.java:
```diff
@@ -0,0 +1,106 @@
+    private Collection<ProgressStep> sequence;
+    private Iterator<ProgressStep> progressIterator;
+
+    ProgressCallbackImpl(FirmwareUpdateHandler firmwareUpdateHanlder, EventPublisher eventPublisher,
```
typo: firmwareUpdateHandler

tomhoefer replied:
thanks

kaikreuzer commented:
@tomhoefer Is it correct that ALL classes are in a public package? Isn't it possible to cleanly split the API/interfaces from internal implementations and services?

Note that so far all interfaces that have to be implemented by binding developers were specifically put in the thing.binding package. Maybe the FirmwareUpdateHandler should go there as well? (note also the imports on the [archetype](https://github.com/eclipse/smarthome/blob/master/tools/archetype/binding/src/main/resources/archetype-resources/META-INF/MANIFEST.MF)).

tomhoefer commented:
Yes... currently all classes are in the public package since all classes are API (execept for ProgressCallbackImpl which is package private)

Of course we can move the two firmware update handler interfaces to the thing.binding package

Furthermore we can think about to introduce an additional package thing.firmware.events which could contain all firmware relatated event stuff

tomhoefer commented:
Updated this PR...

- Fixed the two typos
- Moved all classes / interfaces used by bindings to package org.eclipse.smarthome.core.thing.binding.firmware

Will update the other firmware PRs in the folllowing days

kaikreuzer commented:
Many thanks, lgtm!
